### DISCUSSION TITLE
Move Max Level option to Class Editor

### DISCUSSION THREAD

**Discussion by Guite-Emortal** [AscensionGameDev/Intersect-Engine#1515] [Ideas]
**Is your feature request related to a problem? Please describe.**
I have class tiers and each tier has a different max level cap. There is no way to stall EXP, and even with the use of events to de-level someone once they exceed the level cap they are still rewarded stat points.

**Describe the solution you'd like**
I'd like for you move the Max Level option from the Server Config to the Class editor so that it can be designated per class.

**Comment by Cheshire92** [comment]
I think this sounds like a pretty game specific option. 
Generally speaking most games would have everyone cap out at the same level.

**Comment by Guite-Emortal** [comment]
> I think this sounds like a pretty game specific option.
> Generally speaking most games would have everyone cap out at the same level.

While I can't disagree, that it may be game specific. It doesn't stop people from building their classes with the same max level. It does however open up options for other projects. I draw heavy inspiration from Ragnarok Online, Tree of Savior and other Korean MMO RPGs which do use tiered level caps. I figured there was no harm in asking at the very least.

If it is an absolute No, feel free to close the issue.

**Comment by Cheshire92** [comment]
While people could match the level caps it would also add more work for people to have a default situation just to make a system work for a niche solution. (Unless it would default to the configuration option)

Personally I'd say no, but I figured to keep it open in case anyone else wants to pitch in. It's not a one man show after all.

**Comment by Guite-Emortal** [comment]
I wouldn't mind it defaulting to a pre-configured option. It would most certainly help those who do have a default level cap situation, and only requires minimal adjustment on my end when I raise the cap on a new class.

**Comment by Arufonsu** [comment]
I love mmos like RO, im there with u as my work is focused in such style too 😄, but like ches said, this is way too specific, the normal thing on mmos its a level cap for everyone.

You could always do your own source changes to implement this, this project its open source and you are allowed to do so, good luck with your project!

**Comment by Guite-Emortal** [comment]
> I love mmos like RO, im there with u as my work is focused in such style too 😄, but like ches said, this is way too specific, the normal thing on mmos its a level cap for everyone.
> 
> You could always do your own source changes to implement this, this project its open source and you are allowed to do so, good luck with your project!

If I had the skill to make source changes I wouldn't have asked.

**Comment by Arufonsu** [comment]
> 
> 
> > I love mmos like RO, im there with u as my work is focused in such style too 😄, but like ches said, this is way too specific, the normal thing on mmos its a level cap for everyone.
> > You could always do your own source changes to implement this, this project its open source and you are allowed to do so, good luck with your project!
> 
> If I had the skill to make source changes I wouldn't have asked.

You probably missed the point: the feature is too specific for a engine that aims to reach a general public of game developers.
You are always free to ask for features/enchancements/bug fixes, as long as they fit the current default features / expectations of the devs. It's not up to me, and honestly i'm not sure who's up to decide what stays and what not in this project.

It has nothing to do on whether or not you have the skills to do this, its about the feature itself. You may always pay for it when it comes to something specific related to your own project if you can't do it yourself in this **specific** case, or we can wait until a head dev makes a desition about it 😆

**Comment by jcsnider** [comment]
Decision mostly comes down to me, panda, and Joyce. It's based upon many factors including how much we think it might be used upfront development time required, whether the enhancement justifies the amount of code that must be changed (we do somewhat consider the difficulty of merging the changes for those with custom forks), performance impacts, and finally it comes down to whether or not we are willing to maintain the additional code. 


For the instance of a pet/summoning system:
Would most games use it?
Maybe.

Upfront development required?
Significant, not worth it.

Vastness of code changes?
Pretty wide spread, would impact all combat systems, warping pets would require additional logic, etc

Performance Impacts?
Potentially 2x network packets and processing for players because a pet would be another entity per player.

Long Term Upkeep?
A pain. 


With that thought out a bit, even if someone submitted a PR for pets we probably wouldn't accept it.



For this request specifically I don't know how many games would use it, but I see a few use cases. Development time would be minimal with very few code changes and no real performance impacts. Long term upkeep would also be minimal. So unless there is a big objection I'd probably take this on someday. But maybe not until B8.

**Comment by Cheshire92** [comment]
Assuming we make the max level field default to the configuration setting most people won't even notice the change or ever need to touch it but the option would be there for people that want it.

Sometimes the initial thought is a no but a solution to the reason I'd say no comes up as I write the response. Doesn't seem too difficult or intrusive for people to work with on both sides of the fence.

(Funny thing is, writing this discussion probably already took more time than it would take to make this happen)

**Comment by pandinocoder** [comment]
We might want to change the name from `MaxLevel` to `DefaultMaxLevel` and then have a loading alias so that `MaxLevel` in an (old) configuration doesn't break or get ignored, but any freshly saved configuration file by the server will have `DefaultMaxLevel`.

If it were not clear, I am seconding (thirding?) the idea that the configuration value is used as a default, and classes can just have an override specified.

For the data structure the class should either have a nullable database-stored level property _or_ the non-nullable level property should be 0 for "default". For easy access to the actual class max level, they would also have a non-database-stored level getter that is either `DbNullable ?? Default` or `DbLevel == 0 ? Default : DbLevel` (the names are just pseudocode, use better names).
